package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py

```diff
@@ -503,13 +526,40 @@ def guess_acceptors(self, select='all', max_charge=-0.5):
 
         ag = self.u.select_atoms(select)
         acceptors_ag = ag[ag.charges < max_charge]
-        acceptors_list = np.unique(
-            [
-                '(resname {} and name {})'.format(r, p) for r, p in zip(acceptors_ag.resnames, acceptors_ag.names)
-            ]
-        )
 
-        return " or ".join(acceptors_list)
+        return self.__group_categories(acceptors_ag)
+
+    @staticmethod
+    def __group_categories(group):
+        """ Find categories according to universe constraints
+        
+        Parameters
+        ----------
+        group : AtomGroup
+            AtomGroups corresponding to either hydrogen bond acceptors, 
+            donors, or hydrogen atoms that meant their respective charge
+            and mass constraints. 
+
+        Returns
+        -------
+        select : str
+            String for each hydrogen bond acceptor/donor/hydrogen atom category.
+        """
+
+        if hasattr(group,"resnames") and hasattr(group,"names"):
+            group_list = np.unique(
+                [
+                    '(resname {} and name {})'.format(r, p) for r, p in zip(group.resnames, group.names)
+                ]
+            )
+        else:
```

Comment:
I'm still rather wary of this non optional fallback and also the behaviour this might have on force fields where there are equivalent types of different charges. @MDAnalysis/coredevs can we get a quick look at this please? I'm happy going with the consensus.

Reply:
I'll just point out that if there are no resnames, then a scheme like lammps is most likely used where "equivalent" atoms with different charges are also different types.

Reply:
I appreciate your viewpoint and look forward to others with potential fridge cases. 

I was thinking the fallback behavior allows the current users to perform calculations "business as usual" but would automatically switch to the behavior required for LAMMPS users. Thus, LAMMPS users aren't second class in that the "default" behavior will work for them, rather than requiring them to find a keyword that allows their calculation to run.

Reply:
I don't want this to get too lost, so I'm going to ping the coredev team again @MDAnalysis/coredevs

Reply:
The fact that we're relying on selections here instead of ag instances is iffy. I think that needs to change eventually. Reverse engineering a compact selection from an AG is just rife with potential failures, regardless of using types or resname+name. (We could create an `index` selection but I assume that performance-wise it might be bad to have a selection string with 1M numbers.)

As such I don't think that the proposed change will make the situation worse but it will give users with type-only universes the same functionality. 

I don't think that the problem is really the addition of the alternative *type* selection here. 

----
⚠️ I wanted to suggest to make it
```python
      elif hasattr(group, "types"):
```
and add a final `else` that raises a `NoDataError` but **this point is moot** because [types is a canonical/guessed attribute](https://userguide.mdanalysis.org/stable/topology_system.html#canonical-attributes) and will always be present.

Reply:
Generally agreed with @orbeckst . From skimming the code it looks like:

* the reason we need a selection string is to make an updating atomgroup
* the selection strings output by this method are static? (i.e. name and resname, or, type)

Would it be possible / a good idea to redo the method to return an AtomGroup or UpdatingAtomGroup directly where it's called in `_prepare`, if an `acceptors_sel` is not provided?

(`ag.select_atoms("all", updating=True)`)

Reply:
raise issue to modernize HBA to use AG as objects internally instead of selection strings>?

Reply:
@jaclark5 would you be able to raise this as an issue if not already raised?

Reply:
Thank you for raising #3933.

"Please raise an issue" is a common request in reviews. When we look at code as part of a review and find something needs to be changed but we do not want to overburden you (by requesting it as a change on the current PR) so we ask you to create the TODO. Otherwise this information will get lost. The issue tracker serves as our collective memory.